### DISCUSSION TITLE
`reportUnusedVariable` and `locals()`, variable isn't unused?

### DISCUSSION THREAD

**Discussion by aThorp96** [microsoft/pyright#7345] [Q&A] [answered]
I'm not sure if this is a bug or expected behavior. If a function calls `locals()`, variables which are only accessed via the `dict` that `locals()` returns are considered unused. It's reasonable that Pyright doesn't track all references to dictionary accesses etc so I couldn't expect Pyright to tell me whether or not the variable is actually used. I'm curious though if `locals()` should instead mark a variable as used, since it is in the dictionary after all, or perhaps if a variable is only consumed by `locals()` then it should be marked as _possibly_ unused like [is done in PyLint](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/possibly-unused-variable.html)? 

I won't defend programming like pictured below but it at illustrates the concern: In the first example you can see pyright marks the variable `foo` as an unused variable, effectively marking it as dead code. However in the second example it's clear that removing the unused variable results in an error, because the variable was in use, illustrating that authoritatively calling the variable unused is an error

![Screenshot 2024-02-26 at 12 06 01 PM](https://github.com/microsoft/pyright/assets/28596783/c9dbfe1e-a23e-49c7-9d57-546b8e8e4e8e)

**Comment by erictraut** [comment, marked as answer]
This isn't a bug. Pyright is working as intended here. Python is a very dynamic language, so pyright's diagnostic checks necessarily make some base assumptions about code. If you want your code to work well with static type checkers, avoid using dynamic features like `locals()`.
